### PR TITLE
R1C1 support

### LINKED ISSUE #161: R1C1 notation for formulas
Hello. Is it possible to write formulas in R1C1 notation?

https://learn.microsoft.com/en-us/troubleshoot/microsoft-365-apps/excel/numeric-columns-and-rows#the-r1c1-reference-style

This is useful when you need to programmatically set static references in formulas, instead of recalculating it with every line. ClosedXml [supports](https://github.com/ClosedXML/ClosedXML/wiki/Using-Formulas) thisю

Probably this is also can be added in spreadcheetah? Or probably it's already implemented, and i've just missing it?

### PR DESCRIPTION
**Pull request by SKProCH**
Adds the R1C1 support and tests for it. 

It should cover all the R1C1 use cases.  
Except for the "wrap": RC[-1] in A1 in excel will reference the most right column, wrapping the document. This is currently not supported to not complicate the logic for such rare usecase.

Closes #161

### CODE REVIEW

File: SpreadCheetah/Formulas/R1C1FormulaConverter.cs [hunk 1]
```diff
@@ -0,0 +1,284 @@
...
+        public bool ColumnRelative { get; } = columnRelative;
+        public int Column { get; } = column;
+    }
+
+    /// <summary>
+    /// Converts a formula in the R1C1 reference style to the A1 reference style.
+    /// <paramref name="row"/> and <paramref name="column"/> are the 1-based position of the cell that the formula belongs to,
+    /// which is used as the anchor for relative references.
+    /// </summary>
+    public static string ToA1(string formula, int row, int column)
```
Comment (sveinungf) [inline SpreadCheetah/Formulas/R1C1FormulaConverter.cs:30]: The logic to parse an R1C1 formula is a bit too complex. Can we use Regex instead, and enumerate through matches?

File: SpreadCheetah/Helpers/ThrowHelper.cs [hunk 2]
```diff
@@ -105,6 +105,10 @@ public static void NameTooLong(int maxLength, string? paramName)
     public static void NoteTextTooLong(string? paramName)
         => throw new ArgumentException($"Note text can not exceed {SpreadsheetConstants.MaxNoteTextLength} characters.", paramName);
 
+    [DoesNotReturn]
+    public static void R1C1ReferenceOutOfBounds(string formula)
+        => throw new ArgumentException($"The R1C1 formula '{formula}' contains a reference that is outside the bounds of the worksheet.", nameof(formula));
```
Comment (sveinungf) [inline SpreadCheetah/Helpers/ThrowHelper.cs]: I don't think ArgumentException is the right exception type here, since the argument (the formula) can be valid when used on another cell.

**Review by sveinungf — CHANGES_REQUESTED** [review]

File: SpreadCheetah/Formulas/R1C1FormulaConverter.cs (hunk 1 above)
Comment (SKProCH) [reply SpreadCheetah/Formulas/R1C1FormulaConverter.cs:30]: Yes, of course. But I was thinking that manual algorithm will be more performant than regex one. But if you want - of course.

Comment (sveinungf) [reply SpreadCheetah/Formulas/R1C1FormulaConverter.cs:30]: A manual algorithm can be more performant. But I think it will be difficult to beat the source generated Regex implementation, since it will most likely take advantage of SIMD.

Comment (SKProCH) [reply SpreadCheetah/Formulas/R1C1FormulaConverter.cs:30]: I've did 2 variaants: the original ClosedXml one and optimized one.
I've did an benchmark on test cases data:

image

And even like this I still doen't like how complicated it looks:

image

And slightly optimized variant:

image

Which variant should I use?  
P.S. I still prefer the manual one due to speed, since this is an performance oriented library anyway, and 6x speed diff in potentially hot path

Comment (sveinungf) [reply SpreadCheetah/Formulas/R1C1FormulaConverter.cs:30]: Ouch, that regex pattern is quite complicated. I was hoping it would improve readability, but it is worse than your manual parser. I see it is also several times slower in your benchmark, while I expected the gap to be smaller.

Let's keep your manual parser. You've added a good amount of test cases and you've covered many edge cases so it seems to work correctly. Great work!

**Review by sveinungf — APPROVED** [review]
